# Import Modules and Dataset

In [ ]:
# install modules
!pip install optuna
!pip install autoviz

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 380.1/380.1 kB 5.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 233.4/233.4 kB 7.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 78.8/78.8 kB 6.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 67.5/67.5 kB 1.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 433.8/433.8 kB 7.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.9/1.9 MB 14.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 255.9/255.9 MB 3.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.8/1.8 MB 24.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 24.7/24.7 MB 25.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 11.6/11.6 MB 50.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.0/7.0 MB 56.8 MB/s eta 0:00:00
  Attempting uninstall: xgboost
    Found existing installation: xgboost 2.0.3
    Uninstalling xgboost-2.0.3:
  

In [ ]:
# basic import
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import math

%matplotlib inline

In [ ]:
# data preprocessing
from sklearn.preprocessing import StandardScaler, MinMaxScaler, OneHotEncoder
from sklearn.experimental import enable_iterative_imputer
from sklearn.impute import SimpleImputer, IterativeImputer
from sklearn.cluster import DBSCAN
from sklearn.linear_model import BayesianRidge

In [ ]:
# data visualization
import seaborn as sns
import matplotlib.colors as mat_colors
from mpl_toolkits.mplot3d import Axes3D
from autoviz.AutoViz_Class import AutoViz_Class
from IPython.display import SVG

Imported v0.1.904. Please call AutoViz in this sequence:
    AV = AutoViz_Class()
    %matplotlib inline
    dfte = AV.AutoViz(filename, sep=',', depVar='', dfte=None, header=0, verbose=1, lowess=False,
               chart_format='svg',max_rows_analyzed=150000,max_cols_analyzed=30, save_plot_dir=None)


In [ ]:
# feature selection
from sklearn.feature_selection import SelectKBest, chi2, f_classif
from sklearn.decomposition import PCA, SparsePCA, NMF

In [ ]:
# data modeling
from sklearn.model_selection import train_test_split
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import make_pipeline
from xgboost import XGBClassifier
from lightgbm import LGBMClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier, ExtraTreesClassifier, GradientBoostingClassifier, HistGradientBoostingClassifier
from imblearn.ensemble import BalancedRandomForestClassifier
from sklearn.svm import SVC
from imblearn.over_sampling import SMOTE
from imblearn.pipeline import make_pipeline as make_imb_pipeline
from imblearn.pipeline import make_pipeline as make_pipelines
from imblearn.over_sampling import SMOTE as SMOTEs

In [ ]:
# metric evaluation
from sklearn.metrics import f1_score, classification_report, accuracy_score
from sklearn.model_selection import cross_validate, cross_val_predict, cross_val_score
import sklearn.metrics as metrics

In [ ]:
# hyper parameter and stacking ensemble model
import optuna

In [ ]:
# import data
train_features = pd.read_csv("/data/train_features.csv")
train_labels = pd.read_csv("/data/train_labels.csv")
test_features = pd.read_csv("/data/test_features.csv")

In [ ]:
# concat the train_labels into the train_features
train_features['jumlah_promosi'] = train_labels

In [ ]:
# ID for test_features
ids = test_features['ID']

# Column Overview

**Tentang Dataset**

> Konteks:
Dataset ini memberikan gambaran mendalam tentang detail demografis, status ekonomi, dan perilaku belanja konsumen. Dataset mencakup data penting seperti tahun kelahiran pelanggan, tingkat pendidikan, status pernikahan, pendapatan, struktur keluarga terkait jumlah anak balita dan remaja, serta kebiasaan belanja rinci di berbagai kategori seperti buah, daging, dan ikan. Selain itu, dataset ini juga mencakup metrik perilaku seperti frekuensi pembelian online dan offline, respons terhadap diskon, dan keluhan pelanggan. Dataset ini sangat penting bagi bisnis dan peneliti untuk menganalisis pola perilaku konsumen, sehingga meningkatkan pendekatan pemasaran yang ditargetkan, mengoptimalkan penempatan produk, dan meningkatkan manajemen hubungan pelanggan.

**Glosarium Dataset (Berdasarkan Kolom)**

> `tahun_kelahiran`: Tahun kelahiran pelanggan, memberikan dasar untuk segmentasi demografis dan analisis tren.

> `pendidikan`: Tingkat pendidikan pelanggan, dikategorikan menjadi SMP, SMA, Sarjana, Magister, dan Doktor.

> `status_pernikahan`: Status pernikahan pelanggan, pilihan termasuk Sendiri, Rencana Menikah, Menikah, Cerai, dan Cerai Mati.

> `pendapatan`: Pendapatan tahunan pelanggan dalam Rupiah Indonesia, menunjukkan kelas ekonomi mereka.

> `jumlah_anak_balita`: Jumlah anak balita dalam keluarga pelanggan, relevan untuk memahami kebutuhan pembelian yang berorientasi keluarga.

> `jumlah_anak_remaja`: Jumlah remaja dalam keluarga pelanggan, berguna untuk pemasaran yang ditargetkan pada produk yang menarik bagi konsumen muda.

> `terakhir_belanja`: Hari yang telah berlalu sejak pembelian terakhir pelanggan, metrik untuk mengukur keterlibatan pelanggan dan frekuensi belanja.

> `belanja_buah`: Jumlah yang dihabiskan untuk pembelian buah, mencerminkan preferensi untuk opsi makanan sehat.

> `belanja_daging`: Jumlah yang dihabiskan untuk pembelian daging, menunjukkan kebiasaan diet dan kemungkinan preferensi produk premium.

> `belanja_ikan`: Jumlah yang dihabiskan untuk pembelian ikan, indikator lain dari kebiasaan diet dan kebiasaan belanja untuk barang mudah rusak.

> `belanja_kue`: Pengeluaran untuk item kue, berguna untuk menganalisis perilaku memanjakan diri dan pembelian impuls.

> `pembelian_diskon`: Jumlah pembelian yang dilakukan selama promosi diskon, menunjukkan sensitivitas harga dan respons terhadap penjualan.

> `pembelian_web`: Jumlah pembelian yang dilakukan secara online, memberikan wawasan tentang adopsi platform digital pelanggan untuk berbelanja.

> `pembelian_toko`: Jumlah pembelian yang dilakukan di toko fisik, penting untuk memahami preferensi belanja tradisional.

> `keluhan`: Menunjukkan apakah pelanggan pernah mengajukan keluhan (1 - Ya, 0 - Tidak), ukuran langsung kepuasan pelanggan dan kualitas layanan.

> `tanggal_menjadi_anggota`: Tanggal pelanggan pertama kali terdaftar sebagai anggota, berguna untuk melacak loyalitas dan masa keanggotaan pelanggan.

> `jumlah_promosi`: Jumlah promosi (dari total enam) yang diikuti oleh pelanggan, dengan nilai 0 menunjukkan tidak ada partisipasi, menyoroti tingkat keterlibatan dan efektivitas pemasaran. (TARGET)

# Data Understanding

## Data Overview

In [ ]:
train_features

tahun_kelahiran pendidikan status_pernikahan  pendapatan   \
0          1979         Sarjana    Rencana Menikah          NaN   
1          1950         Sarjana    Rencana Menikah   84063000.0   
2          1966         Sarjana            Menikah  127532564.0   
3          1961        Magister    Rencana Menikah  165579620.0   
4          1970         Sarjana    Rencana Menikah  117703159.0   
...               ...        ...               ...          ...   
3812       1955        Magister            Menikah   78199470.0   
3813       1947          Doktor    Rencana Menikah  109306000.0   
3814       1974        Magister            Menikah  104621000.0   
3815       1957             SMA    Rencana Menikah  110850000.0   
3816       1961         Sarjana            Menikah  130512000.0   

      jumlah_anak_balita  jumlah_anak_remaja  terakhir_belanja  belanja_buah  \
0             0.0                 1.0                NaN           50575.0     
1             NaN                 NaN               70.0            6069.0     
2             0.0                 0.0               45.0          117611.0     
3             0.0                 0.0               90.0          206346.0     
4             1.0                 1.0               78.0           90563.0     
...                  ...                 ...               ...           ...   
3812          0.0                 0.0               33.0            6069.0     
3813          0.0                 1.0               44.0               0.0     
3814          0.0                 2.0               68.0            2023.0     
3815          1.0                 1.0               67.0           18207.0     
3816          0.0                 0.0               70.0          329749.0     

      belanja_daging  belanja_ikan  belanja_kue  pembelian_diskon  \
0         260967.0       50575.0      20230.0           2.0         
1          44506.0       80920.0      20230.0           9.0         
2         265460.0       96341.0     145573.0           1.0         
3        1613901.0       27725.0     125868.0           0.0         
4         311757.0       40358.0      33875.0           7.0         
...              ...           ...          ...               ...   
3812       25977.0        3856.0       5784.0           5.0         
3813       50575.0           NaN          0.0           3.0         
3814       62713.0        8092.0          0.0           7.0         
3815       70805.0       24276.0          NaN           4.0         
3816      600831.0      420784.0     161840.0           3.0         

      pembelian_web  pembelian_toko  keluhan tanggal_menjadi_anggota  \
0          2.0             5.0         0.0          2014-05-05         
1          6.0             4.0         0.0          2013-03-17         
2          1.0             7.0         0.0                 NaN         
3          7.0             8.0         0.0                 NaN         
4          6.0             5.0         0.0                 NaN         
...             ...             ...      ...                     ...   
3812       1.0             0.0         0.0                 NaN         
3813       6.0             3.0         0.0          2014-06-09         
3814       5.0             7.0         0.0          2013-11-07         
3815       5.0             4.0         0.0          2013-06-30         
3816       7.0            15.0         0.0          2014-01-25         

      jumlah_promosi  
0            2        
1            0        
2            1        
3            4        
4            4        
...              ...  
3812         5        
3813         1        
3814         0        
3815         0        
3816         0        

[3817 rows x 17 columns]

In [ ]:
test_features

ID   tahun_kelahiran pendidikan status_pernikahan  pendapatan   \
0     2241       1957         Sarjana                NaN  120660151.0   
1     2274       1968          Doktor            Menikah  163551821.0   
2     1107       1968             SMA            Menikah   29857000.0   
3     4478       1971          Doktor            Menikah  117949098.0   
4     5080       1974         Sarjana    Rencana Menikah  164761134.0   
...    ...              ...        ...               ...          ...   
3813   271       1984        Magister            Sendiri   66320000.0   
3814  3334       1972         Sarjana            Menikah  159651946.0   
3815  3478       1968        Magister    Rencana Menikah   84058923.0   
3816    89       1966         Sarjana    Rencana Menikah  106301000.0   
3817  2151       1973         Sarjana            Menikah  140353000.0   

      jumlah_anak_balita  jumlah_anak_remaja  terakhir_belanja  belanja_buah  \
0             0.0                 1.0               63.0          122277.0     
1             0.0                 1.0               58.0           35761.0     
2             0.0                 0.0               34.0            8092.0     
3             0.0                 1.0               82.0            4872.0     
4             0.0                 0.0               28.0          343208.0     
...                  ...                 ...               ...           ...   
3813          1.0                 0.0               19.0           12138.0     
3814          0.0                 0.0               90.0           32119.0     
3815          0.0                 1.0               46.0            8198.0     
3816          0.0                 1.0               74.0           16184.0     
3817          0.0                 0.0                9.0           62713.0     

      belanja_daging  belanja_ikan  belanja_kue  pembelian_diskon  \
0         541399.0      214192.0      84305.0           4.0         
1         353335.0       63365.0      41112.0           2.0         
2          22253.0       30345.0      26299.0           2.0         
3         126061.0           0.0       9745.0           3.0         
4        1416462.0      236196.0     107776.0           0.0         
...              ...           ...          ...               ...   
3813       38437.0           0.0      10115.0           4.0         
3814     1032587.0      176078.0      78693.0           0.0         
3815       47595.0       11392.0      10540.0           NaN         
3816       78897.0           NaN          0.0           3.0         
3817     1707412.0      453152.0          NaN           2.0         

      pembelian_web  pembelian_toko  keluhan tanggal_menjadi_anggota  
0          6.0            10.0         0.0                 NaN        
1          5.0            10.0         0.0                 NaN        
2          0.0             5.0         0.0          2013-08-06        
3          5.0             7.0         0.0                 NaN        
4          1.0             8.0         0.0                 NaN        
...             ...             ...      ...                     ...  
3813       0.0             6.0         1.0          2014-03-17        
3814       6.0             8.0         0.0                 NaN        
3815       0.0             1.0         0.0                 NaN        
3816       5.0             3.0         0.0          2014-01-20        
3817       2.0             8.0         0.0          2012-09-15        

[3818 rows x 17 columns]

In [ ]:
train_labels

jumlah_promosi
0            2      
1            0      
2            1      
3            4      
4            4      
...              ...
3812         5      
3813         1      
3814         0      
3815         0      
3816         0      

[3817 rows x 1 columns]

In [ ]:
train_features.shape

(3817, 17)

In [ ]:
train_features.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3817 entries, 0 to 3816
Data columns (total 17 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   tahun_kelahiran          3817 non-null   int64  
 1   pendidikan               3628 non-null   object 
 2   status_pernikahan        3605 non-null   object 
 3   pendapatan               3627 non-null   float64
 4   jumlah_anak_balita       3627 non-null   float64
 5   jumlah_anak_remaja       3613 non-null   float64
 6   terakhir_belanja         3645 non-null   float64
 7   belanja_buah             3636 non-null   float64
 8   belanja_daging           3639 non-null   float64
 9   belanja_ikan             3624 non-null   float64
 10  belanja_kue              3603 non-null   float64
 11  pembelian_diskon         3639 non-null   float64
 12  pembelian_web            3652 non-null   float64
 13  pembelian_toko           3648 non-null   float64
 14  keluhan                 

In [ ]:
test_features.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3818 entries, 0 to 3817
Data columns (total 17 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   ID                       3818 non-null   int64  
 1   tahun_kelahiran          3818 non-null   int64  
 2   pendidikan               3614 non-null   object 
 3   status_pernikahan        3636 non-null   object 
 4   pendapatan               3615 non-null   float64
 5   jumlah_anak_balita       3609 non-null   float64
 6   jumlah_anak_remaja       3608 non-null   float64
 7   terakhir_belanja         3617 non-null   float64
 8   belanja_buah             3632 non-null   float64
 9   belanja_daging           3623 non-null   float64
 10  belanja_ikan             3622 non-null   float64
 11  belanja_kue              3631 non-null   float64
 12  pembelian_diskon         3639 non-null   float64
 13  pembelian_web            3638 non-null   float64
 14  pembelian_toko          

In [ ]:
train_features.describe()

tahun_kelahiran   pendapatan   jumlah_anak_balita  jumlah_anak_remaja  \
count    3817.000000    3.627000e+03      3627.000000         3613.000000      
mean     1967.823946    1.144832e+08         0.293080            0.353723      
std        11.768131    4.346042e+07         0.473063            0.493014      
min      1899.000000    5.073000e+06         0.000000            0.000000      
25%      1959.000000    8.112512e+07         0.000000            0.000000      
50%      1968.000000    1.156214e+08         0.000000            0.000000      
75%      1976.000000    1.504960e+08         1.000000            1.000000      
max      2000.000000    3.328840e+08         2.000000            2.000000      

       terakhir_belanja  belanja_buah   belanja_daging  belanja_ikan   \
count     3645.000000      3636.000000   3.639000e+03     3624.000000   
mean        47.231550     59804.239824   4.385748e+05    81428.997792   
std         27.068512     74024.976109   5.120427e+05    99976.226855   
min          0.000000         0.000000   0.000000e+00        0.000000   
25%         25.000000      7907.000000   4.947950e+04    10115.000000   
50%         47.000000     26456.000000   2.219930e+05    36054.500000   
75%         69.000000     86162.000000   6.863555e+05   121380.000000   
max        128.000000    396508.000000   3.489675e+06   621600.000000   

        belanja_kue   pembelian_diskon  pembelian_web  pembelian_toko  \
count    3603.000000     3639.000000     3652.000000     3648.000000    
mean    63377.970580        2.125584        4.436473        5.767818    
std     79435.457282        2.100133        3.002522        3.210738    
min         0.000000        0.000000        0.000000        0.000000    
25%      7947.000000        0.000000        2.000000        3.000000    
50%     27795.000000        2.000000        4.000000        5.000000    
75%     89502.500000        3.000000        7.000000        8.000000    
max    542164.000000       20.000000       30.000000       17.000000    

         keluhan    jumlah_promosi  
count  3621.000000    3817.000000   
mean      0.004971       2.585800   
std       0.070340       2.101845   
min       0.000000       0.000000   
25%       0.000000       0.000000   
50%       0.000000       3.000000   
75%       0.000000       4.000000   
max       1.000000       6.000000

In [ ]:
train_features.describe(include="all")

tahun_kelahiran pendidikan status_pernikahan   pendapatan   \
count     3817.000000        3628               3605  3.627000e+03   
unique            NaN           6                  6           NaN   
top               NaN     Sarjana    Rencana Menikah           NaN   
freq              NaN        1841               1271           NaN   
mean      1967.823946         NaN                NaN  1.144832e+08   
std         11.768131         NaN                NaN  4.346042e+07   
min       1899.000000         NaN                NaN  5.073000e+06   
25%       1959.000000         NaN                NaN  8.112512e+07   
50%       1968.000000         NaN                NaN  1.156214e+08   
75%       1976.000000         NaN                NaN  1.504960e+08   
max       2000.000000         NaN                NaN  3.328840e+08   

        jumlah_anak_balita  jumlah_anak_remaja  terakhir_belanja  \
count       3627.000000         3613.000000        3645.000000     
unique              NaN                 NaN                NaN     
top                 NaN                 NaN                NaN     
freq                NaN                 NaN                NaN     
mean           0.293080            0.353723          47.231550     
std            0.473063            0.493014          27.068512     
min            0.000000            0.000000           0.000000     
25%            0.000000            0.000000          25.000000     
50%            0.000000            0.000000          47.000000     
75%            1.000000            1.000000          69.000000     
max            2.000000            2.000000         128.000000     

        belanja_buah   belanja_daging  belanja_ikan    belanja_kue   \
count     3636.000000   3.639000e+03     3624.000000    3603.000000   
unique            NaN            NaN             NaN            NaN   
top               NaN            NaN             NaN            NaN   
freq              NaN            NaN             NaN            NaN   
mean     59804.239824   4.385748e+05    81428.997792   63377.970580   
std      74024.976109   5.120427e+05    99976.226855   79435.457282   
min          0.000000   0.000000e+00        0.000000       0.000000   
25%       7907.000000   4.947950e+04    10115.000000    7947.000000   
50%      26456.000000   2.219930e+05    36054.500000   27795.000000   
75%      86162.000000   6.863555e+05   121380.000000   89502.500000   
max     396508.000000   3.489675e+06   621600.000000  542164.000000   

        pembelian_diskon  pembelian_web  pembelian_toko    keluhan    \
count      3639.000000     3652.000000     3648.000000   3621.000000   
unique             NaN             NaN             NaN           NaN   
top                NaN             NaN             NaN           NaN   
freq               NaN             NaN             NaN           NaN   
mean          2.125584        4.436473        5.767818      0.004971   
std           2.100133        3.002522        3.210738      0.070340   
min           0.000000        0.000000        0.000000      0.000000   
25%           0.000000        2.000000        3.000000      0.000000   
50%           2.000000        4.000000        5.000000      0.000000   
75%           3.000000        7.000000        8.000000      0.000000   
max          20.000000       30.000000       17.000000      1.000000   

       tanggal_menjadi_anggota  jumlah_promosi  
count               1065          3817.000000   
unique               564                  NaN   
top           2013-05-17                  NaN   
freq                   6                  NaN   
mean                 NaN             2.585800   
std                  NaN             2.101845   
min                  NaN             0.000000   
25%                  NaN             0.000000   
50%                  NaN             3.000000   
75%                  NaN             4.000000   
max                  NaN             6.000000

In [ ]:
test_features.describe()

ID       tahun_kelahiran   pendapatan   jumlah_anak_balita  \
count  3818.000000    3818.000000    3.615000e+03      3609.000000      
mean   3841.507072    1967.789942    1.163249e+08         0.282904      
std    2198.401006      11.808572    5.105122e+07         0.465005      
min       2.000000    1892.000000    2.998899e+06         0.000000      
25%    1950.750000    1959.000000    8.168162e+07         0.000000      
50%    3872.500000    1968.000000    1.177321e+08         0.000000      
75%    5770.750000    1976.000000    1.506877e+08         1.000000      
max    7634.000000    2009.000000    1.305740e+09         2.000000      

       jumlah_anak_remaja  terakhir_belanja  belanja_buah   belanja_daging  \
count      3608.000000        3617.000000      3632.000000   3.623000e+03    
mean          0.355322          47.181366     61234.083425   4.528484e+05    
std           0.496865          26.940287     76647.590617   5.123206e+05    
min           0.000000           0.000000         0.000000   0.000000e+00    
25%           0.000000          25.000000      7497.500000   5.462100e+04    
50%           0.000000          46.000000     29525.000000   2.327530e+05    
75%           1.000000          68.000000     86776.750000   7.217960e+05    
max           2.000000         133.000000    408646.000000   3.271191e+06    

       belanja_ikan    belanja_kue   pembelian_diskon  pembelian_web  \
count    3622.000000    3631.000000     3639.000000     3638.000000    
mean    85338.612093   63801.505921        2.101676        4.465091    
std    103789.546199   78428.544713        2.115571        2.935615    
min         0.000000       0.000000        0.000000        0.000000    
25%     10115.000000    8092.000000        0.000000        2.000000    
50%     40025.500000   29251.000000        2.000000        4.000000    
75%    129656.750000   91407.000000        3.000000        6.000000    
max    574932.000000  532049.000000       17.000000       27.000000    

       pembelian_toko    keluhan    
count    3632.000000   3625.000000  
mean        5.835628      0.002759  
std         3.288349      0.052457  
min         0.000000      0.000000  
25%         3.000000      0.000000  
50%         5.000000      0.000000  
75%         8.000000      0.000000  
max        17.000000      1.000000

In [ ]:
test_features.describe(include="all")

ID       tahun_kelahiran pendidikan status_pernikahan  \
count   3818.000000    3818.000000        3614               3636   
unique          NaN            NaN           6                  6   
top             NaN            NaN     Sarjana    Rencana Menikah   
freq            NaN            NaN        1935               1272   
mean    3841.507072    1967.789942         NaN                NaN   
std     2198.401006      11.808572         NaN                NaN   
min        2.000000    1892.000000         NaN                NaN   
25%     1950.750000    1959.000000         NaN                NaN   
50%     3872.500000    1968.000000         NaN                NaN   
75%     5770.750000    1976.000000         NaN                NaN   
max     7634.000000    2009.000000         NaN                NaN   

         pendapatan   jumlah_anak_balita  jumlah_anak_remaja  \
count   3.615000e+03      3609.000000         3608.000000      
unique           NaN              NaN                 NaN      
top              NaN              NaN                 NaN      
freq             NaN              NaN                 NaN      
mean    1.163249e+08         0.282904            0.355322      
std     5.105122e+07         0.465005            0.496865      
min     2.998899e+06         0.000000            0.000000      
25%     8.168162e+07         0.000000            0.000000      
50%     1.177321e+08         0.000000            0.000000      
75%     1.506877e+08         1.000000            1.000000      
max     1.305740e+09         2.000000            2.000000      

        terakhir_belanja  belanja_buah   belanja_daging  belanja_ikan   \
count      3617.000000      3632.000000   3.623000e+03     3622.000000   
unique             NaN              NaN            NaN             NaN   
top                NaN              NaN            NaN             NaN   
freq               NaN              NaN            NaN             NaN   
mean         47.181366     61234.083425   4.528484e+05    85338.612093   
std          26.940287     76647.590617   5.123206e+05   103789.546199   
min           0.000000         0.000000   0.000000e+00        0.000000   
25%          25.000000      7497.500000   5.462100e+04    10115.000000   
50%          46.000000     29525.000000   2.327530e+05    40025.500000   
75%          68.000000     86776.750000   7.217960e+05   129656.750000   
max         133.000000    408646.000000   3.271191e+06   574932.000000   

         belanja_kue   pembelian_diskon  pembelian_web  pembelian_toko  \
count     3631.000000     3639.000000     3638.000000     3632.000000    
unique            NaN             NaN             NaN             NaN    
top               NaN             NaN             NaN             NaN    
freq              NaN             NaN             NaN             NaN    
mean     63801.505921        2.101676        4.465091        5.835628    
std      78428.544713        2.115571        2.935615        3.288349    
min          0.000000        0.000000        0.000000        0.000000    
25%       8092.000000        0.000000        2.000000        3.000000    
50%      29251.000000        2.000000        4.000000        5.000000    
75%      91407.000000        3.000000        6.000000        8.000000    
max     532049.000000       17.000000       27.000000       17.000000    

          keluhan   tanggal_menjadi_anggota  
count   3625.000000              1059        
unique          NaN               545        
top             NaN        2013-08-06        
freq            NaN                 6        
mean       0.002759               NaN        
std        0.052457               NaN        
min        0.000000               NaN        
25%        0.000000               NaN        
50%        0.000000               NaN        
75%        0.000000               NaN        
max        1.000000               NaN

## Value Counts Analysis

In [ ]:
def show_value_count(data, label):
  value_count = data[label].value_counts()

  print(f"\nPersebaran dari kolom {label}: \n{value_count}")

In [ ]:
column_types = train_features.dtypes
object_columns = column_types[column_types == 'object'].index.tolist()

object_columns

['pendidikan', 'status_pernikahan', 'tanggal_menjadi_anggota']

In [ ]:
integer_columns = column_types[column_types == 'int64'].index.tolist()
integer_columns

['tahun_kelahiran', 'jumlah_promosi']

In [ ]:
float_columns = column_types[column_types == 'float64'].index.tolist()
float_columns

['pendapatan',
 'jumlah_anak_balita',
 'jumlah_anak_remaja',
 'terakhir_belanja',
 'belanja_buah',
 'belanja_daging',
 'belanja_ikan',
 'belanja_kue',
 'pembelian_diskon',
 'pembelian_web',
 'pembelian_toko',
 'keluhan']

In [ ]:
for column in object_columns:
  show_value_count(train_features, column)


Persebaran dari kolom pendidikan: 
pendidikan
Sarjana     1841
Magister     856
SMA          435
Doktor       403
SMP           91
5              2
Name: count, dtype: int64

Persebaran dari kolom status_pernikahan: 
status_pernikahan
Rencana Menikah    1271
Menikah            1114
Sendiri             965
Cerai               214
Cerai Mati           40
5                     1
Name: count, dtype: int64

Persebaran dari kolom tanggal_menjadi_anggota: 
tanggal_menjadi_anggota
2013-05-17    6
2013-09-25    6
2013-07-03    6
2013-01-10    6
2013-01-16    5
             ..
2014-02-01    1
2013-10-10    1
2013-06-18    1
2013-02-04    1
2014-06-09    1
Name: count, Length: 564, dtype: int64


In [ ]:
for column in object_columns:
  show_value_count(test_features, column)


Persebaran dari kolom pendidikan: 
pendidikan
Sarjana     1935
Magister     803
SMA          408
Doktor       404
SMP           62
5              2
Name: count, dtype: int64

Persebaran dari kolom status_pernikahan: 
status_pernikahan
Rencana Menikah    1272
Menikah            1126
Sendiri             972
Cerai               206
Cerai Mati           59
5                     1
Name: count, dtype: int64

Persebaran dari kolom tanggal_menjadi_anggota: 
tanggal_menjadi_anggota
2013-08-06    6
2014-04-03    6
2013-09-01    5
2013-04-20    5
2012-11-07    5
             ..
2012-11-16    1
2012-10-29    1
2013-05-18    1
2013-01-09    1
2013-04-18    1
Name: count, Length: 545, dtype: int64


In [ ]:
for column in float_columns:
  show_value_count(train_features, column)


Persebaran dari kolom pendapatan: 
pendapatan
58555000.0     2
98033000.0     2
100016000.0    2
102971000.0    2
147845000.0    2
              ..
75153084.0     1
43279588.0     1
263538206.0    1
112547078.0    1
130512000.0    1
Name: count, Length: 3622, dtype: int64

Persebaran dari kolom jumlah_anak_balita: 
jumlah_anak_balita
0.0    2594
1.0    1003
2.0      30
Name: count, dtype: int64

Persebaran dari kolom jumlah_anak_remaja: 
jumlah_anak_remaja
0.0    2361
1.0    1226
2.0      26
Name: count, dtype: int64

Persebaran dari kolom terakhir_belanja: 
terakhir_belanja
48.0     57
39.0     55
23.0     53
27.0     52
25.0     50
30.0     50
67.0     49
19.0     49
47.0     48
55.0     48
0.0      47
63.0     46
8.0      46
41.0     45
53.0     45
33.0     45
71.0     44
59.0     44
60.0     44
70.0     44
74.0     43
11.0     43
46.0     43
20.0     43
61.0     42
16.0     42
66.0     42
68.0     41
58.0     41
32.0     40
73.0     40
17.0     40
9.0      40
62.0     40
38.0     

In [ ]:
train_features[train_features["keluhan"] == 1.0]

tahun_kelahiran pendidikan status_pernikahan  pendapatan   \
262        1984        Magister            Sendiri   66319917.0   
981        1951         Sarjana    Rencana Menikah   98962000.0   
992        1952         Sarjana              Cerai   43427000.0   
1090       1986             SMA            Menikah   31357000.0   
1325       1960         Sarjana    Rencana Menikah  115264000.0   
1566       1987        Magister            Sendiri          NaN   
1567       1982             SMP    Rencana Menikah   52692736.0   
1872       1958         Sarjana            Menikah  114782000.0   
2134       1975         Sarjana            Sendiri   75793000.0   
2172       1959         Sarjana            Sendiri  166101000.0   
2599       1987        Magister            Sendiri   66360252.0   
2698       1982         Sarjana    Rencana Menikah   81639000.0   
2890       1981          Doktor            Sendiri   83951000.0   
3041       1984             SMA            Menikah   31299799.0   
3154       1975         Sarjana            Menikah  131001000.0   
3268       1954         Sarjana            Menikah   76683000.0   
3272       1984             SMP            Sendiri   52690222.0   
3632       1980         Sarjana            Menikah   36087000.0   

      jumlah_anak_balita  jumlah_anak_remaja  terakhir_belanja  belanja_buah  \
262           1.0                 0.0               18.0           12305.0     
981           0.0                 1.0               36.0           26299.0     
992           1.0                 1.0               34.0               NaN     
1090          1.0                 0.0                4.0           10115.0     
1325          0.0                 1.0               23.0          125426.0     
1566          1.0                 0.0               15.0           12534.0     
1567          0.0                 0.0                5.0               NaN     
1872          0.0                 1.0               69.0          208369.0     
2134          1.0                 1.0               46.0            2023.0     
2172          1.0                 1.0               45.0          271082.0     
2599          1.0                 0.0                6.0           10643.0     
2698          1.0                 0.0               53.0            6069.0     
2890          1.0                 NaN               39.0               0.0     
3041          1.0                 0.0               19.0               NaN     
3154          0.0                 0.0               89.0           86989.0     
3268          1.0                 1.0               94.0               0.0     
3272          0.0                 0.0                8.0            1018.0     
3632          1.0                 0.0               83.0               0.0     

      belanja_daging  belanja_ikan  belanja_kue  pembelian_diskon  \
262       38938.0            0.0      10226.0           4.0         
981      159817.0        12138.0      24276.0           5.0         
992           0.0            NaN      10115.0           0.0         
1090      60690.0        14161.0       4046.0           5.0         
1325     352002.0       157794.0          0.0           8.0         
1566      38437.0         1586.0       9123.0           4.0         
1567       7595.0         4372.0      18807.0           0.0         
1872     333795.0       208369.0     157794.0           4.0         
2134      38437.0            0.0          0.0           2.0         
2172     811223.0       204323.0      42483.0           5.0         
2599      35075.0         2988.0      10862.0           4.0         
2698      26299.0            0.0          0.0           1.0         
2890      64736.0        16184.0      12138.0           2.0         
3041      69506.0        10303.0        913.0           6.0         
3154     774809.0        32368.0      14161.0           1.0         
3268      26299.0         4046.0          0.0           1.0         
3272        665.0         2715.0      

In [ ]:
for column in integer_columns:
  show_value_count(train_features, column)


Persebaran dari kolom tahun_kelahiran: 
tahun_kelahiran
1968    140
1970    137
1969    134
1972    130
1973    125
1960    114
1965    114
1959    113
1977    108
1967    108
1962    108
1974    107
1961    107
1957    105
1971    105
1966    103
1975    101
1963    101
1964     96
1976     94
1979     94
1958     93
1956     84
1980     83
1978     78
1954     75
1955     72
1983     68
1953     66
1981     66
1984     65
1952     63
1951     62
1985     57
1982     56
1988     47
1986     44
1948     42
1989     40
1950     40
1949     39
1987     38
1947     28
1990     25
1945     20
1946     20
1991     19
1944     11
1994     10
1943      9
1993      8
1992      7
1996      5
1942      4
1941      3
1995      3
1902      2
1999      2
1940      2
1920      1
1914      1
1900      1
1924      1
1932      1
1929      1
1997      1
1909      1
1918      1
2000      1
1922      1
1899      1
1998      1
1916      1
1912      1
1925      1
1908      1
Name: count, dtype: int64

Pers

## Missing Data

In [ ]:
train_features.isna().sum()

tahun_kelahiran               0
pendidikan                  189
status_pernikahan           212
pendapatan                  190
jumlah_anak_balita          190
jumlah_anak_remaja          204
terakhir_belanja            172
belanja_buah                181
belanja_daging              178
belanja_ikan                193
belanja_kue                 214
pembelian_diskon            178
pembelian_web               165
pembelian_toko              169
keluhan                     196
tanggal_menjadi_anggota    2752
jumlah_promosi                0
dtype: int64

In [ ]:
test_features.isna().sum()

ID                            0
tahun_kelahiran               0
pendidikan                  204
status_pernikahan           182
pendapatan                  203
jumlah_anak_balita          209
jumlah_anak_remaja          210
terakhir_belanja            201
belanja_buah                186
belanja_daging              195
belanja_ikan                196
belanja_kue                 187
pembelian_diskon            179
pembelian_web               180
pembelian_toko              186
keluhan                     193
tanggal_menjadi_anggota    2759
dtype: int64

## Duplicate Data

In [ ]:
train_features.duplicated().sum()

0

In [ ]:
test_features.duplicated().sum()

0

# Pre Visualization + EDA

## Null Values Imputer

### Imputer Mode

In [ ]:
train_features['pendidikan'] = train_features['pendidikan'].fillna(train_features['pendidikan'].mode()[0])
test_features['pendidikan'] = test_features['pendidikan'].fillna(train_features['pendidikan'].mode()[0])

train_features['pendidikan'] = train_features['pendidikan'].fillna(train_features['pendidikan'].mode()[0])
test_features['pendidikan'] = test_features['pendidikan'].fillna(train_features['pendidikan'].mode()[0])

train_features['status_pernikahan'] = train_features['status_pernikahan'].fillna(train_features['status_pernikahan'].mode()[0])
test_features['status_pernikahan'] = test_features['status_pernikahan'].fillna(train_features['status_pernikahan'].mode()[0])

In [ ]:
train_features['jumlah_anak_balita'] = train_features['jumlah_anak_balita'].fillna(train_features['jumlah_anak_balita'].mode()[0])
test_features['jumlah_anak_balita'] = test_features['jumlah_anak_balita'].fillna(train_features['jumlah_anak_balita'].mode()[0])

train_features['jumlah_anak_remaja'] = train_features['jumlah_anak_remaja'].fillna(train_features['jumlah_anak_remaja'].mode()[0])
test_features['jumlah_anak_remaja'] = test_features['jumlah_anak_remaja'].fillna(train_features['jumlah_anak_remaja'].mode()[0])

train_features['keluhan'] = train_features['keluhan'].fillna(train_features['keluhan'].mode()[0])
test_features['keluhan'] = test_features['keluhan'].fillna(train_features['keluhan'].mode()[0])

### Imputer Median

In [ ]:
imputer_median = SimpleImputer(strategy='median')
imputer_median.fit(train_features[['pendapatan', 'belanja_buah', 'belanja_daging', 'belanja_ikan', 'belanja_kue']])

train_features[['pendapatan', 'belanja_buah', 'belanja_daging', 'belanja_ikan', 'belanja_kue']] = imputer_median.transform(train_features[['pendapatan', 'belanja_buah', 'belanja_daging', 'belanja_ikan', 'belanja_kue']])
test_features[['pendapatan', 'belanja_buah', 'belanja_daging', 'belanja_ikan', 'belanja_kue']] = imputer_median.transform(test_features[['pendapatan', 'belanja_buah', 'belanja_daging', 'belanja_ikan', 'belanja_kue']])

imputer_terakhir_belanja = SimpleImputer(strategy='median')
imputer_terakhir_belanja.fit(train_features[['terakhir_belanja']])

train_features['terakhir_belanja'] = imputer_terakhir_belanja.transform(train_features[['terakhir_belanja']])
test_features['terakhir_belanja'] = imputer_terakhir_belanja.transform(test_features[['terakhir_belanja']])

imputer_pembelian = SimpleImputer(strategy='median')
imputer_pembelian.fit(train_features[['pembelian_diskon', 'pembelian_web', 'pembelian_toko']])

train_features[['pembelian_diskon', 'pembelian_web', 'pembelian_toko']] = imputer_pembelian.transform(train_features[['pembelian_diskon', 'pembelian_web', 'pembelian_toko']])
test_features[['pembelian_diskon', 'pembelian_web', 'pembelian_toko']] = imputer_pembelian.transform(test_features[['pembelian_diskon', 'pembelian_web', 'pembelian_toko']])

## '5' Values into Mode

In [ ]:
train_features.loc[train_features['pendidikan'] == '5', 'pendidikan'] = train_features['pendidikan'].mode()[0]
train_features.loc[train_features['status_pernikahan'] == '5', 'status_pernikahan'] = train_features['status_pernikahan'].mode()[0]

test_features.loc[test_features['pendidikan'] == '5', 'pendidikan'] = test_features['pendidikan'].mode()[0]
test_features.loc[test_features['status_pernikahan'] == '5', 'status_pernikahan'] = test_features['status_pernikahan'].mode()[0]

## Encoding

In [ ]:
# 1. encode 'pendidikan' and 'status_pernikahan' manually
train_features['pendidikan_encoded'] = train_features['pendidikan'].map({'SMP': 0, 'SMA': 1, 'Sarjana': 2, 'Magister': 3, 'Doktor': 4})
test_features['pendidikan_encoded'] = test_features['pendidikan'].map({'SMP': 0, 'SMA': 1, 'Sarjana': 2, 'Magister': 3, 'Doktor': 4})

# drop original columns
train_features.drop(columns=['pendidikan'], inplace=True)
test_features.drop(columns=['pendidikan'], inplace=True)

# One hot encoding
train_features = pd.get_dummies(train_features, columns= ['status_pernikahan'])
test_features = pd.get_dummies(test_features, columns= ['status_pernikahan'])

## Frequency and Monetary (FM)

In [ ]:
train_features['total_pembelian'] = train_features['pembelian_diskon'] + train_features['pembelian_web'] + train_features['pembelian_toko']
train_features['total_belanja'] = train_features['belanja_buah'] + train_features['belanja_daging'] + train_features['belanja_ikan'] + train_features['belanja_kue']

test_features['total_pembelian'] = test_features['pembelian_diskon'] + test_features['pembelian_web'] + test_features['pembelian_toko']
test_features['total_belanja'] = test_features['belanja_buah'] + test_features['belanja_daging'] + test_features['belanja_ikan'] + test_features['belanja_kue']

## Capping Outlier

In [ ]:
# Replace values less than 1940 with 1940
train_features.loc[train_features['tahun_kelahiran'] < 1940, 'tahun_kelahiran'] = 1940
test_features.loc[test_features['tahun_kelahiran'] < 1940, 'tahun_kelahiran'] = 1940

# Replace values more than 2000 with 2000
train_features.loc[train_features['tahun_kelahiran'] > 2000, 'tahun_kelahiran'] = 2000
test_features.loc[test_features['tahun_kelahiran'] > 2000, 'tahun_kelahiran'] = 2000

## Column Engineer

### Member Status Column

In [ ]:
train_features["status_member"] = train_features['tanggal_menjadi_anggota'].apply(lambda x: 0 if x == 'missing' else 1)
test_features["status_member"] = test_features['tanggal_menjadi_anggota'].apply(lambda x: 0 if x == 'missing' else 1)

### Pendapatan by Family Column

In [ ]:
# Calculate pendapatan_by_family, filling with 0 if the denominator is 0
train_features["pendapatan_by_family"] = np.where(train_features['jumlah_anak_balita'] + train_features['jumlah_anak_remaja'] == 0, 0, train_features['pendapatan'] / (train_features['jumlah_anak_balita'] + train_features['jumlah_anak_remaja']))
# Repeat the same process for test_features
test_features["pendapatan_by_family"] = np.where(test_features['jumlah_anak_balita'] + test_features['jumlah_anak_remaja'] == 0, 0, test_features['pendapatan'] / (test_features['jumlah_anak_balita'] + test_features['jumlah_anak_remaja']))
# train_features["pendapatan_by_family"].fillna(0, inplace=True)

### Umur (Based on 2014) Column

In [ ]:
train_features['umur'] = train_features.tahun_kelahiran.apply(lambda x: 2014 - int(x))
test_features['umur'] = test_features.tahun_kelahiran.apply(lambda x: 2014 - int(x))

### Pendapatan by Anak Remaja Column

In [ ]:
# Calculate pendapatan_by_anak_remaja, filling with 0 if the denominator is 0
train_features["pendapatan_by_anak_remaja"] = np.where(train_features['jumlah_anak_remaja'] == 0, 0, train_features['pendapatan'] / train_features['jumlah_anak_remaja'])

# Repeat the same process for test_features
test_features["pendapatan_by_anak_remaja"] = np.where(test_features['jumlah_anak_remaja'] == 0, 0, test_features['pendapatan'] / test_features['jumlah_anak_remaja'])

### Age Grouping Column

In [ ]:
# Divide people in the 5 age group

# 10-16: 0
# 17-30: 1
# 31-45: 2
# 46-60: 3
# 60+: 4

train_features['kelompok_umur'] = pd.cut(train_features.umur,bins=[1,17,30,45,60,140],labels=['0','1','2','3','4'])
test_features['kelompok_umur'] = pd.cut(test_features.umur,bins=[1,17,30,45,60,140],labels=['0','1','2','3','4'])

In [ ]:
# 0-24: 0
# 25-49: 1
# 50-75: 2
# 75-100: 3

# Calculate percentiles of the income distribution
percentiles = np.percentile(train_features['pendapatan'], [0, 25, 50, 75, 100])

# Print the calculated percentiles
print("15th Percentile:", percentiles[0])
print("25th Percentile:", percentiles[1])
print("35th Percentile:", percentiles[2])
print("45th Percentile:", percentiles[3])
print("55th Percentile:", percentiles[4])
# print("65th Percentile (Median):", percentiles[5])
# print("75th Percentile:", percentiles[6])

15th Percentile: 5073000.0
25th Percentile: 82608533.0
35th Percentile: 115621394.0
45th Percentile: 147953453.0
55th Percentile: 332884000.0


### Pendapatan Grouping Column

In [ ]:
train_features['kelompok_pendapatan'] = pd.cut(train_features.pendapatan,bins=[0, 2998000.0,5000000.0,82608533.0,115621394.0,147953453.0,333884000.0,1305740000.0],labels=['0','1','2','3','4','5','6'])
test_features['kelompok_pendapatan'] = pd.cut(test_features.pendapatan,bins=[0, 2998000.0,5000000.0,82608533.0,115621394.0,147953453.0,333884000.0, 1305740000.0],labels=['0','1','2','3','4','5','6'])

In [ ]:
train_features['pendapatan'].min(), test_features['pendapatan'].min()

(5073000.0, 2998899.0)

In [ ]:
train_features['pendapatan'].max(), test_features['pendapatan'].max()

(332884000.0, 1305740000.0)

In [ ]:
train_features.isna().sum(), test_features.isna().sum()

(tahun_kelahiran                         0
 pendapatan                              0
 jumlah_anak_balita                      0
 jumlah_anak_remaja                      0
 terakhir_belanja                        0
 belanja_buah                            0
 belanja_daging                          0
 belanja_ikan                            0
 belanja_kue                             0
 pembelian_diskon                        0
 pembelian_web                           0
 pembelian_toko                          0
 keluhan                                 0
 tanggal_menjadi_anggota              2752
 jumlah_promosi                          0
 pendidikan_encoded                      0
 status_pernikahan_Cerai                 0
 status_pernikahan_Cerai Mati            0
 status_pernikahan_Menikah               0
 status_pernikahan_Rencana Menikah       0
 status_pernikahan_Sendiri               0
 total_pembelian                         0
 total_belanja                           0
 status_mem

### Total Belanja by Pendapatan Column

In [ ]:
train_features['total_belanja_by_pendapatan'] = train_features['total_belanja']/train_features['pendapatan']
test_features['total_belanja_by_pendapatan'] = test_features['total_belanja']/test_features['pendapatan']

## Fixing Weird Column 'category' Behavior

In [ ]:
# Convert categorical columns to integers
train_features['kelompok_umur'] = train_features['kelompok_umur'].astype(int)
test_features['kelompok_umur'] = test_features['kelompok_umur'].astype(int)

train_features['kelompok_pendapatan'] = train_features['kelompok_pendapatan'].astype(int)
test_features['kelompok_pendapatan'] = test_features['kelompok_pendapatan'].astype(int)

## Drop Useless Column

In [ ]:
train_features.drop(columns='keluhan',inplace=True)
test_features.drop(columns='keluhan',inplace=True)

train_features.drop(columns='tanggal_menjadi_anggota',inplace=True)
test_features.drop(columns='tanggal_menjadi_anggota',inplace=True)

## Column Update

In [ ]:
column_types = train_features.dtypes
object_columns = column_types[column_types == 'object'].index.tolist()

object_columns

[]

In [ ]:
integer_columns = column_types[column_types == 'int64'].index.tolist()
integer_columns

['tahun_kelahiran',
 'jumlah_promosi',
 'pendidikan_encoded',
 'status_member',
 'umur',
 'kelompok_umur',
 'kelompok_pendapatan']

In [ ]:
float_columns = column_types[column_types == 'float64'].index.tolist()
float_columns

['pendapatan',
 'jumlah_anak_balita',
 'jumlah_anak_remaja',
 'terakhir_belanja',
 'belanja_buah',
 'belanja_daging',
 'belanja_ikan',
 'belanja_kue',
 'pembelian_diskon',
 'pembelian_web',
 'pembelian_toko',
 'total_pembelian',
 'total_belanja',
 'pendapatan_by_family',
 'pendapatan_by_anak_remaja',
 'total_belanja_by_pendapatan']

# EDA Analysis

In [ ]:
df = train_features

enol = df[df['jumlah_promosi'] == 0].describe().T
satu = df[df['jumlah_promosi'] == 1].describe().T
dua = df[df['jumlah_promosi'] == 2].describe().T
tiga = df[df['jumlah_promosi'] == 3].describe().T
empat = df[df['jumlah_promosi'] == 4].describe().T
lima = df[df['jumlah_promosi'] == 5].describe().T
enam = df[df['jumlah_promosi'] == 6].describe().T

# Menggabungkan semua statistik deskriptif menjadi satu DataFrame
df_combined = pd.concat([enol['mean'], satu['mean'], dua['mean'], tiga['mean'], empat['mean'], lima['mean'], enam['mean']], axis=1)
df_combined.columns = ['Enol Promosi', 'Satu Promosi', 'Dua Promosi', 'Tiga Promosi', 'Empat Promosi', 'Lima Promosi', 'Enam Promosi']

# Membuat heatmap dari DataFrame yang telah digabungkan
plt.figure(figsize=(12, 6))
sns.heatmap(df_combined, annot=True, cmap='Blues', linewidths=0.4, linecolor='black', fmt='.2f')
plt.title('Mean: Masing-masing Jumlah Promosi')
plt.show()

## Bar Plot

### Mengeck persebaran jumlah anak balita dan jumlah anak remaja suatu customer (Berdasarkan Pendidikan dan Status Pernikahan)

In [ ]:
for column in object_columns:
  if column == 'tanggal_menjadi_anggota':
      continue

  # Menghitung frekuensi setiap kategori
  order = train_features[column].value_counts().index

  plt.figure(figsize=(24, 8))  # Lebih besar untuk menampung kedua plot

  # Plot untuk anak balita
  plt.subplot(1, 2, 1)  # Satu baris, dua kolom, plot pertama
  ax1 = sns.countplot(x=column, data=train_features, hue='jumlah_anak_balita', order=order, palette="Set1")
  plt.title(f'Distribusi {column.title()} (Berdasarkan Jumlah Anak Balita)')
  plt.xlabel(f'Kategori {column.title()}')
  plt.ylabel('Jumlah Anak Balita')
  # Menambahkan label jumlah pada bar
  for p in ax1.patches:
      ax1.annotate(f'{int(p.get_height())}', (p.get_x() + p.get_width() / 2., p.get_height()),
                    ha='center', va='center', fontsize=10, color='black', xytext=(0, 5),
                    textcoords='offset points')

  # Plot untuk anak remaja
  plt.subplot(1, 2, 2)  # Satu baris, dua kolom, plot kedua
  ax2 = sns.countplot(x=column, data=train_features, hue='jumlah_anak_remaja', order=order, palette="Set1")
  plt.title(f'Distribusi {column.title()} (Berdasarkan Jumlah Anak Remaja)')
  plt.xlabel(f'Kategori {column.title()}')
  plt.ylabel('Jumlah Anak Remaja')
  # Menambahkan label jumlah pada bar
  for p in ax2.patches:
      ax2.annotate(f'{int(p.get_height())}', (p.get_x() + p.get_width() / 2., p.get_height()),
                    ha='center', va='center', fontsize=10, color='black', xytext=(0, 5),
                    textcoords='offset points')

  plt.tight_layout()
  plt.show()

Conclusion :
- Untuk kategori pendidikan tinggi (Sarjana, Magister, Doktor) terlihat mereka cenderung mempunyai remaja dibandingkan balita. Sedangkan pendidikan mnengah (SMA, SMP) sebaliknya.
- SMP 1/2 memiliki 0 balita dan 1/2 memiliki 1 balita, berbeda dengan jenjang lainnya.
- NaN menunjukan sifat pendidikan menengah dimana cenderung mempunyai balita daripada remaja.

- Untuk kategori Pernikahan, selain Sendiri customer cenderung memiliki remaja dibandingkan balita minimal 1. Sedangkan mayoritas customer balita 2 lebih sering dibandingkan memiliki 2 remaja dalam satu keluarga.

### Persebaran jumlah_promosi

In [ ]:
dict_hues = {
    "jumlah_anak_balita":"jumlah_anak_remaja",
}

In [ ]:
for k, v in dict_hues.items():
  column = "jumlah_promosi"
  if column == 'tanggal_menjadi_anggota':
      break

  plt.figure(figsize=(24, 8))  # Lebih besar untuk menampung kedua plot

  # Plot untuk anak balita
  plt.subplot(1, 2, 1)  # Satu baris, dua kolom, plot pertama
  ax1 = sns.countplot(x=column, data=train_features, hue=k, palette="Set1")
  plt.title(f'Distribusi {column.title()} (Berdasarkan {k.title()})')
  plt.xlabel(f'Kategori {column.title()}')
  plt.ylabel(k.title())
  # Menambahkan label jumlah pada bar
  for p in ax1.patches:
      ax1.annotate(f'{int(p.get_height())}', (p.get_x() + p.get_width() / 2., p.get_height()),
                    ha='center', va='center', fontsize=10, color='black', xytext=(0, 5),
                    textcoords='offset points')

  # Plot untuk anak remaja
  plt.subplot(1, 2, 2)  # Satu baris, dua kolom, plot kedua
  ax2 = sns.countplot(x=column, data=train_features, hue=v, palette="Set1")
  plt.title(f'Distribusi {column.title()} (Berdasarkan {v.title()})')
  plt.xlabel(f'Kategori {column.title()}')
  plt.ylabel(v.title())
  # Menambahkan label jumlah pada bar
  for p in ax2.patches:
      ax2.annotate(f'{int(p.get_height())}', (p.get_x() + p.get_width() / 2., p.get_height()),
                    ha='center', va='center', fontsize=10, color='black', xytext=(0, 5),
                    textcoords='offset points')

  plt.tight_layout()
  plt.show()

## Box Plot

In [ ]:
def boxplot_method(df, int_column, title=""):

  plt.figure(figsize=(6, 6))
  sns.set(style="darkgrid")
  ax = sns.boxplot(x=int_column, data=df, palette="Set1")
  plt.title(title, fontsize=16)
  plt.show()

### Melihat Persebaran kolom numerical untuk melihat outlier

In [ ]:
for column in integer_columns+float_columns:
  boxplot_method(train_features, column, f"Box Plot {column}")

### Analisis Outlier Tahun Kelahiran

In [ ]:
boxplot_method(train_features, 'tahun_kelahiran', f"Box Plot Tahun Kelahiran (Train)")
boxplot_method(test_features, 'tahun_kelahiran', f"Box Plot Tahun Kelahiran (Test)")

In [ ]:
kelahiran2000 = train_features[train_features["tahun_kelahiran"] == 2000]
kelahiran2000

In [ ]:
kelahiran2000_test = test_features[test_features["tahun_kelahiran"] >= 2000]
kelahiran2000_test

In [ ]:
kelahiran1940 = train_features[train_features["tahun_kelahiran"] < 1940]

In [ ]:
kelahiran1940_test = test_features[test_features["tahun_kelahiran"] < 1940]
kelahiran1940_test

In [ ]:
sns.set_theme(style="darkgrid")
sns.histplot(data=train_features, x="tahun_kelahiran")
plt.show()

### Analisis Outlier Pendapatan

In [ ]:
boxplot_method(train_features, 'pendapatan', f"Box Plot pendapatan (Train)")
boxplot_method(test_features, 'pendapatan', f"Box Plot pendapatan (Test)")

In [ ]:
pendapatan_outlier = train_features[train_features["pendapatan"] > 2.5 * 10 ** 8]
pendapatan_outlier

In [ ]:
pendapatan_outlier_test = test_features[test_features["pendapatan"] > 2.5 * 10 ** 8]
pendapatan_outlier_test

### Analisis outlier belanja

In [ ]:
boxplot_method(train_features, 'belanja_buah', f"Box Plot belanja (Train)")
boxplot_method(test_features, 'belanja_buah', f"Box Plot belanja (Test)")

In [ ]:
train_features[train_features["belanja_buah"] > 2.5 * 10 ** 8]

### Analisis Outlier pembelian

In [ ]:
boxplot_method(train_features, 'pembelian_toko', f"Box Plot pembelian (Train)")
boxplot_method(test_features, 'pembelian_toko', f"Box Plot pembelian (Test)")

In [ ]:
train_features[train_features["pembelian_toko"] > 15]

In [ ]:
test_features[test_features["pembelian_toko"] > 15]

In [ ]:
train_features[train_features['total_belanja'] == 0.0].iloc[:, :20]

In [ ]:
train_features[train_features['pembelian_diskon'] < (train_features['pembelian_web']+train_features['pembelian_toko'])].iloc[:, :20]

In [ ]:
test_features[(test_features['pembelian_toko'].isna()) & (test_features['pembelian_web'].isna())]

In [ ]:
test_features[(test_features['pembelian_toko'].isna()) & (test_features['pembelian_diskon'].isna())]

## Histogram

In [ ]:
def show_histogram(df):
  cols = 3
  # Calculate the number of rows needed
  rows = math.ceil(len(float_columns + integer_columns) / cols)

  # Create subplot
  fig, axs = plt.subplots(rows, cols, figsize=(5 * cols, 5 * rows))
  # Flatten the array of axes, so it's easier to iterate over
  axs = axs.flatten()

  # Loop over the list of columns
  for i, col in enumerate(float_columns + integer_columns):
      sns.histplot(df[col], ax=axs[i], kde=True, color='royalblue', element='bars', stat='count')
      axs[i].set_title(f'Distribution of {col}', fontsize=12)
      axs[i].set_xlabel('')
      axs[i].set_ylabel('Frequency')
      axs[i].grid(True)

  # Adjust layout
  plt.tight_layout()
  plt.show()

In [ ]:
show_histogram(train_features)

In [ ]:
def show_histogram_test(df):
  cols = 3
  # Calculate the number of rows needed
  rows = math.ceil(len(float_columns + integer_columns) / cols)

  # Create subplot
  fig, axs = plt.subplots(rows, cols, figsize=(5 * cols, 5 * rows))
  # Flatten the array of axes, so it's easier to iterate over
  axs = axs.flatten()

  # Loop over the list of columns
  for i, col in enumerate(float_columns + integer_columns):
    if 'promosi' in col or 'churn' in col:
      continue
    sns.histplot(df[col], ax=axs[i], kde=True, color='royalblue', element='bars', stat='count')
    axs[i].set_title(f'Distribution of {col}', fontsize=12)
    axs[i].set_xlabel('')
    axs[i].set_ylabel('Frequency')
    axs[i].grid(True)

  # Adjust layout
  plt.tight_layout()
  plt.show()

show_histogram_test(test_features)

## Pair Plot

### Analisis RFM (Recency, Frequency, Monetary) Sum

In [ ]:
sns.pairplot(train_features[['pendapatan', 'terakhir_belanja', 'total_belanja', 'total_pembelian', 'jumlah_promosi']], hue='jumlah_promosi', palette='deep')
plt.show()

### Analisis pembelian dengan pendapatan

In [ ]:
sns.pairplot(train_features[['pendapatan', 'jumlah_promosi']+[column for column in train_features.columns if 'pembelian' in column]], hue='jumlah_promosi', palette='deep')
plt.show()

### Analisis belanja dengan pendapatan

In [ ]:
sns.pairplot(train_features[['pendapatan', 'jumlah_promosi']+[column for column in train_features.columns if 'belanja' in column]], hue='jumlah_promosi' , palette='deep')
plt.show()

## Pie Plot

In [ ]:
for column in object_columns + ['jumlah_promosi']:
    if column == 'tanggal_menjadi_anggota':
        continue
    # Make sure the data is in a suitable format for Pandas Series operations
    if isinstance(train_features[column], list):
        pie_data = pd.Series(train_features[column])
    else:
        pie_data = train_features[column]  # Assuming this is already a Series or similar

    value_counts = pie_data.value_counts()  # Get the count of each unique value

    unique_values = len(value_counts)
    colors = plt.cm.tab20.colors[:unique_values]  # Select distinct colors

    # Create the pie chart
    plt.figure(figsize=(6, 6))  # Set a larger figure size for clarity
    plt.pie(value_counts, labels=[f'{idx}' for idx, val in value_counts.items()], colors=colors,
            autopct=lambda p: f'{p:.1f}%' if p > 0 else '', startangle=90)  # Use a lambda to format the percentages
    plt.title(f'Komposisi Kolom "{column}"\nDengan {unique_values} Unique Values', fontsize=16)
    plt.axis('equal')  # Ensure pie is drawn as a circle
    plt.tight_layout()  # Adjust subplots to give the pie chart some space

    plt.show()

## Heat Map Correlation

In [ ]:
plt.figure(figsize=(30,30))
sns.heatmap(train_features.select_dtypes(include=[np.number]).corr(), annot=True);
plt.show()

In [ ]:
plt.figure(figsize=(30,30))
sns.heatmap(test_features.select_dtypes(include=[np.number]).corr(), annot=True);
plt.show()

# AutoViz Analysis

## Prepare Visual

In [ ]:
from autoviz.AutoViz_Class import AutoViz_Class

AV = AutoViz_Class()

In [ ]:
train_features.to_csv('train_features.csv', index=False)
filename = '/content/train_features.csv'  # The path to your dataset
sep = ","

AV.AutoViz(filename="",
           sep=',',
           depVar='jumlah_promosi',
           dfte=train_features,
           header=0,
           verbose=2,
           lowess=False,
           chart_format='svg',
           max_rows_analyzed=4000,
           max_cols_analyzed=40
)

## Bar Plots

In [ ]:
SVG(filename='/content/AutoViz_Plots/jumlah_promosi/Bar_Plots.svg')

## Box Plots

In [ ]:
SVG(filename='/content/AutoViz_Plots/jumlah_promosi/Box_Plots.svg')

## Distribution Plots (Numerical Column)

In [ ]:
SVG(filename='/content/AutoViz_Plots/jumlah_promosi/Dist_Plots_Numerics.svg')

## Distribution Plots (Target Column)

In [ ]:
SVG(filename='/content/AutoViz_Plots/jumlah_promosi/Dist_Plots_target.svg')

## Heat Maps

In [ ]:
SVG(filename='/content/AutoViz_Plots/jumlah_promosi/Heat_Maps.svg')

## Pair Scatter Plots

In [ ]:
SVG(filename='/content/AutoViz_Plots/jumlah_promosi/Pair_Scatter_Plots.svg')

## Scatter Plots

In [ ]:
SVG(filename='/content/AutoViz_Plots/jumlah_promosi/Scatter_Plots.svg')

## Time Series Plot

In [ ]:
SVG(filename='/content/AutoViz_Plots/jumlah_promosi/Time_Series_Plots.svg')

# Modeling

## Prepare Model

In [ ]:
# menyamakan kolom pada train dan test features
test_features = test_features.reindex(columns=train_features.drop('jumlah_promosi', axis=1).columns, fill_value=0)

In [ ]:
X = train_features.drop('jumlah_promosi', axis=1)
y = train_labels.squeeze()
# X = X.drop(['pendapatan','total_belanja','belanja_daging'],axis = 1)
# test_features = test_features.drop(['pendapatan','total_belanja','belanja_daging'],axis = 1)

In [ ]:
preprocessor = 'passthrough'

X.shape, y.shape

## Explore Model

In [ ]:
SEED = 42

models = [
    ('Decision Tree', DecisionTreeClassifier()),
    ('Random Forest', RandomForestClassifier()),
    ('et', ExtraTreesClassifier()),
    ('SVM', SVC()),
    ('xgb', XGBClassifier()),
    ('gb', GradientBoostingClassifier()),
    ('lgbm', LGBMClassifier()),
    ('hgb', HistGradientBoostingClassifier()),
    ('brf', BalancedRandomForestClassifier())
]

In [ ]:
for clf_name, clf in models:
    pipeline = make_pipelines(preprocessor, SMOTEs(random_state=42), clf)

    y_pred = cross_val_predict(pipeline, X, y, cv=5)

    print(f"Classification Report for {clf_name}:")
    print(classification_report(y, y_pred))
    print(metrics.confusion_matrix(y, y_pred))

# Metric Evaluation

In [ ]:
SEED = 101

# Define the parameters
best_params = {'max_depth': 10,
               'learning_rate': 0.022128449020639294,
               'n_estimators': 960,
               'subsample': 0.7452516398232679,
               'colsample_bytree': 0.6733222527016066,
               'min_child_weight': 0.12023717394916922}

models = [
    # ('xgb', XGBClassifier(**best_params)),
    # # ('Decision Tree', DecisionTreeClassifier()),
    # ('Random Forest', RandomForestClassifier()),
    # # ('SVM', SVC()),
    # ('rf', RandomForestClassifier(**best_params)),
    # ('et', ExtraTreesClassifier(random_state=SEED)),
    ('xgb', XGBClassifier(**best_params)),
    # # ('gb', GradientBoostingClassifier(random_state=SEED)),
    ('lgb', LGBMClassifier(**best_params)),
    # ('hgb', HistGradientBoostingClassifier(**best_params)),
    # ('rf_en', RandomForestClassifier(criterion='entropy')),
    # ('brf', BalancedRandomForestClassifier())
]

In [ ]:
import sklearn.metrics as metrics

for clf_name, clf in models:
    pipeline = make_pipelines(preprocessor, SMOTEs(random_state=42), clf)

    y_pred = cross_val_predict(pipeline, X, y, cv=5)

    print(f"Classification Report for {clf_name}:")
    print(classification_report(y, y_pred))
    print(metrics.confusion_matrix(y, y_pred))

# Hyper Tuning + Parameter

In [ ]:
import optuna
from sklearn.model_selection import cross_val_predict
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
import sklearn.metrics as metrics

def objective(trial):
    # Define hyperparameter search space
    xgb_params = {
        'max_depth': trial.suggest_int('max_depth', 3, 10),
        'learning_rate': trial.suggest_float('learning_rate', 0.001, 0.1, log=True),
        'n_estimators': trial.suggest_int('n_estimators', 100, 1000),
        'subsample': trial.suggest_float('subsample', 0.5, 1.0),
        'colsample_bytree': trial.suggest_float('colsample_bytree', 0.5, 1.0),
        'min_child_weight': trial.suggest_float('min_child_weight', 0.1, 10.0, log=True),
    }

    # Create XGBoost classifier with suggested hyperparameters
    clf = XGBClassifier(**xgb_params)

    # Create pipeline with preprocessing, oversampling, and classifier
    pipeline = make_pipeline(preprocessor, SMOTE(random_state=42), clf)

    # Perform cross-validation and return negative accuracy (to maximize)
    y_pred = cross_val_predict(pipeline, X, y, cv=5)
    accuracy = accuracy_score(y, y_pred)
    return accuracy

# Run Optuna optimization
study = optuna.create_study(direction='maximize')
study.optimize(objective, n_trials=100)

# Get best hyperparameters and their corresponding accuracy
best_params = study.best_params
best_accuracy = study.best_value

# Print best hyperparameters and accuracy
print("Best Parameters:", best_params)
print("Best Accuracy:", best_accuracy)

# Create final XGBoost classifier with best hyperparameters
best_clf = XGBClassifier(**best_params)
best_pipeline = make_pipeline(preprocessor, SMOTE(random_state=42), best_clf)

# Perform cross-validation with best classifier and print classification report
y_pred = cross_val_predict(best_pipeline, X, y, cv=5)
print("Classification Report:")
print(classification_report(y, y_pred))

Best Parameters: {'max_depth': 10, 'learning_rate': 0.0361149792012908, 'n_estimators': 885, 'subsample': 0.7522753681495064, 'colsample_bytree': 0.5828268875128704, 'min_child_weight': 0.14970337564484035}
Best Accuracy: 0.7471836520827875
Classification Report:
              precision    recall  f1-score   support

           0       0.64      0.75      0.69       983
           1       0.54      0.42      0.47       515
           2       0.74      0.67      0.71       389
           3       0.87      0.85      0.86       472
           4       0.87      0.90      0.89       515
           5       0.86      0.88      0.87       557
           6       0.79      0.75      0.77       386

    accuracy                           0.75      3817
   macro avg       0.76      0.74      0.75      3817
weighted avg       0.74      0.75      0.74      3817


# Export Result

In [ ]:
tuning_parameter = {'max_depth': 10,
                    'learning_rate': 0.0361149792012908,
                    'n_estimators': 885,
                    'subsample': 0.7522753681495064,
                    'colsample_bytree': 0.5828268875128704,
                    'min_child_weight': 0.14970337564484035
}

In [ ]:
# Create final XGBoost classifier with best hyperparameters
best_clf = XGBClassifier(**tuning_parameter)
best_pipeline = make_pipelines(preprocessor, SMOTEs(random_state=42), best_clf)

# Perform cross-validation with best classifier and print classification report
y_pred = cross_val_predict(best_pipeline, X, y, cv=5)
print("Classification Report:")
print(classification_report(y, y_pred))

In [ ]:
best_clf.fit(X, y)

# predict test dan bikin csv-nya
test_predictions = best_clf.predict(test_features)

submission_xgb = pd.DataFrame({
    'ID': ids,
    'jumlah_promosi': test_predictions
})

submission_xgb.to_csv('submission.csv', index=False)